File: src/grid-calculator.ts
export default class GridCalculator {
  private _width: number
  private _height: number

  constructor(width: number, height: number) {
    this._width = width
    this._height = height
  }

  public width(): number {
    return this._width
  }

  public height(): number {
    return this._height
  }

  public getLength(): number {
    return this._width * this._height
  }

  public getColumnOfIndex(index: number): number {
    return index % this._width
  }

  public getRowOfIndex(index: number): number {
    return Math.floor(index / this._width)
  }

  public getIndex(row: number, col: number): number {
    return (this._width * row) + col;
  }
}

File: golang/grid-calculator.go
package main

import "math"

type GridCalculator struct {
  width int
  height int
}

func (g GridCalculator) GetLength() int {
  return g.width * g.height;
}

func (g GridCalculator) GetColumnFromIndex(index int) int {
  return index % g.width;
}

func (g GridCalculator) GetRowFromIndex(index int) int {
  return int(math.Floor(float64(index / g.width)));
}

func (g GridCalculator) GetIndex(row int, col int) int {
  return (g.width * row) + col;
}
